### DISCUSSION TITLE
Maven supports compilation and execution on classpath or module path

### DISCUSSION THREAD

**Discussion by hendrikebbers** [support-and-care/maven-support-and-care#35] [Ideas for Funding]
This is based on the discussion in https://github.com/OpenElements/maven-support-care/discussions/11

We need to refactor Maven core + a hand full of plugins (compiler, surefire, exec...) to be able to compilation and execution modules and applications on classpath or module path.

As a next step, we need to define the important plugins that need to be refactored and estimate the needed resources.

**Comment by hendrikebbers** [comment]
Next to that, all discussions for possible changes that are needed to fulfill the goal (like moving code to maven core) should happen here.

**Comment by rmannibucau** [comment]
Likely more embracing new type usage for that concern since feature is there since day one of jpms at a bit verbosity cost so no new feature, more a refacto.

**Comment by slawekjaranowski** [comment]
https://issues.apache.org/jira/browse/MNG-8015 - is it not address such issue?

**Reply by hendrikebbers** [reply]
I assume so at least for Maven core. No idea what additional changes are needed in compile, surefire, exec ... plugins

**Reply by hendrikebbers** [reply]
@desruisseaux has the best overview and I hope that he will add some detail here in future

**Reply by desruisseaux** [reply]
Yes, MNG-8015 allows users to specify their intend. This part has been already delivered in Maven core. The next step is to modify the plugins for taking those intend in account. Currently, all plugins ignore what the users have said.

**Comment by gnodet** [comment]
Plugins should be moved one by one to the maven 4 API which will provide support for enhanced class/module paths.

**Comment by ascheman** [comment]
This discussion has been converted to Epic #49 for tracking.

Closing this discussion.
